src/main/scala/mesosphere/marathon/core/externalvolume/impl/providers/DVDIProvider.scala

```diff
@@ -20,91 +20,87 @@ import scala.collection.immutable.Set
-
-      if (offset >= 0) (offset + 1).toString else ""
-    }
-
-    def mkVar(name: String, value: String): Environment.Variable =
-      Environment.Variable.newBuilder.setName(name).setValue(value).build
-
-    val vars = Seq[Environment.Variable](
-      mkVar(dvdiVolumeContainerPath + suffix, vol.containerPath),
-      mkVar(dvdiVolumeName + suffix, vol.external.name),
-      mkVar(dvdiVolumeDriver + suffix, vol.external.options(driverOption))
-    )
-
-    val optsVar = {
+    def dockerVolumeParameters(volume: ExternalVolume): Seq[Parameter] = {
+      import OptionLabelPatterns._
       val prefix: String = name + OptionNamespaceSeparator
       // don't let the user override these
       val ignore = Set(driverOption)
       // external.size trumps any user-specified dvdi/size option
-      val opts = vol.external.options ++ Map[String, String](
-        vol.external.size.map(prefix + "size" -> _.toString).toList: _*
+      val opts = volume.external.options ++ Map[String, String](
+        volume.external.size.map(prefix + "size" -> _.toString).toList: _*
       )
 
       // forward all dvdi/* options to the dvdcli driver, stripping the dvdi/ prefix
       // and trimming the values
       opts.filterKeys{ k =>
         k.startsWith(prefix) && !ignore.contains(k.toLowerCase)
       }.map{
-        case (k, v) => k.substring(prefix.length) + "=" + v.trim()
-      }.mkString(",")
+        case (k, v) => Parameter.newBuilder
+          .setKey(k.substring(prefix.length))
+          .setValue(v.trim())
+          .build
+      }.toSeq
+    }
+
+    def applyOptions(dv: MesosDockerVolumeBuilder, opts: Seq[Parameter]): MesosDockerVolume = {
+      if (opts.isEmpty) {
+        dv.clearDriverOptions
+      }
+      else {
+        dv.setDriverOptions(Parameters.newBuilder.addAllParameter(opts.asJava))
+      }
+      dv.build
     }
 
-    if (optsVar.isEmpty) vars
-    else { vars :+ mkVar(dvdiVolumeOpts + suffix, optsVar) }
+    def toUnifiedMesosVolume(volume: ExternalVolume): MesosVolume = {
+      val driverName = volume.external.options(driverOption)
+      MesosVolume.newBuilder
+        .setContainerPath(volume.containerPath)
+        .setMode(volume.mode)
+        .setSource(MesosVolume.Source.newBuilder
+          .setType(MesosVolume.Source.Type.DOCKER_VOLUME)
+          .setDockerVolume(applyOptions(
+            MesosVolume.Source.DockerVolume.newBuilder
+              .setDriver(driverName)
+              .setName(volume.external.name),
+            dockerVolumeParameters(volume))
+          )
+        ).build
+    }
+  } // Builders
+
+  override def build(builder: ContainerInfo.Builder, ev: ExternalVolume): Unit = {
+    // special behavior for docker vs. mesos containers
+    // - docker containerizer: serialize volumes into mesos proto
+    // - docker containerizer: specify "volumeDriver" for the container
+    if (builder.getType == ContainerInfo.Type.DOCKER && builder.hasDocker) {
```

Comment:
What about `builder.getType == ContainerInfo.Type.DOCKER && !builder.hasDocker` ? Should we create the docker section?